File: tests/python/test_maaq.py
"""
Python binding tests for MAAQ indicator.
These tests mirror the Rust unit tests to ensure Python bindings work correctly.
"""
import pytest
import numpy as np
from test_utils import load_test_data, assert_close, EXPECTED_OUTPUTS
from rust_comparison import compare_with_rust


try:
    from my_project import (
        maaq, 
        maaq_batch,
        MaaqStream
    )
except ImportError:
    pytest.skip("MAAQ module not available - run 'maturin develop' first", allow_module_level=True)


class TestMaaq:
    """Test suite for MAAQ indicator Python bindings"""
    
    @pytest.fixture(scope='class')
    def test_data(self):
        return load_test_data()
    
    def test_maaq_partial_params(self, test_data):
        """Test MAAQ with default parameters - mirrors check_maaq_partial_params"""
        close = np.array(test_data['close'], dtype=np.float64)
        
        
        result = maaq(close, 11, 2, 30)
        assert len(result) == len(close)


    def test_maaq_accuracy(self, test_data):
        """Test MAAQ matches expected values from Rust tests - mirrors check_maaq_accuracy"""
        close = np.array(test_data['close'], dtype=np.float64)
        expected = EXPECTED_OUTPUTS['maaq']
        
        
        result = maaq(
            close,
            period=expected['default_params']['period'],
            fast_period=expected['default_params']['fast_period'],
            slow_period=expected['default_params']['slow_period']
        )
        
        assert len(result) == len(close)
        
        
        assert_close(
            result[-5:], 
            expected['last_5_values'], 
            atol=1e-2,
            msg="MAAQ last 5 values mismatch"
        )
        
        
        


    def test_maaq_zero_period(self):
        """Test MAAQ fails with zero period - mirrors check_maaq_zero_period"""
        input_data = np.array([10.0, 20.0, 30.0], dtype=np.float64)
        
        
        with pytest.raises(ValueError, match="maaq:"):
            maaq(input_data, 0, 2, 30)
        
        
        with pytest.raises(ValueError, match="maaq:"):
            maaq(input_data, 11, 0, 30)
        
        
        with pytest.raises(ValueError, match="maaq:"):
            maaq(input_data, 11, 2, 0)


    def test_maaq_period_exceeds_length(self):
        """Test MAAQ fails with period exceeding data length - mirrors check_maaq_period_exceeds_length"""
        data_small = np.array([10.0, 20.0, 30.0], dtype=np.float64)
        
        with pytest.raises(ValueError, match="maaq:"):
            maaq(data_small, 10, 2, 30)


    def test_maaq_very_small_dataset(self):
        """Test MAAQ fails with insufficient data - mirrors check_maaq_very_small_dataset"""
        single_point = np.array([42.0], dtype=np.float64)
        
        with pytest.raises(ValueError, match="maaq:"):
            maaq(single_point, 11, 2, 30)


    def test_maaq_empty_input(self):
        """Test MAAQ with empty input"""
        data_empty = np.array([], dtype=np.float64)
        
        with pytest.raises(ValueError, match="maaq:"):
            maaq(data_empty, 11, 2, 30)


    def test_maaq_all_nan(self):
        """Test MAAQ with all NaN input"""
        data = np.array([np.nan, np.nan, np.nan, np.nan, np.nan], dtype=np.float64)
        
        with pytest.raises(ValueError, match="maaq:"):
            maaq(data, 3, 2, 5)


    def test_maaq_reinput(self, test_data):
        """Test MAAQ with re-input of MAAQ result - mirrors check_maaq_reinput"""
        close = np.array(test_data['close'], dtype=np.float64)
        
        
        first_result = maaq(close, 11, 2, 30)
        
        
        second_result = maaq(first_result, 20, 3, 25)
        
        assert len(second_result) == len(first_result)
        
        
        
        for i in range(40, len(second_result)):
            assert np.isfinite(second_result[i]), f"Unexpected NaN at index {i}"


    def test_maaq_nan_handling(self, test_data):
        """Test MAAQ handling of NaN values - mirrors check_maaq_nan_handling
        
        Verifies that MAAQ properly creates NaN values during warmup period
        and produces valid values after warmup.
        """
        close = np.array(test_data['close'], dtype=np.float64)
    
        
        period = 11
        result = maaq(close, period, 2, 30)
        
        assert len(result) == len(close)
        
        
        for i in range(period - 1):
            assert np.isnan(result[i]), f"Expected NaN at warmup index {i}, got {result[i]}"
        
        
        
        if len(result) > 240:
            for i in range(240, len(result)):
                assert not np.isnan(result[i]), f"Found unexpected NaN at index {i}"


    def test_maaq_batch(self, test_data):
        """Test MAAQ batch computation"""
        close = np.array(test_data['close'], dtype=np.float64)
    
        
        batch_result = maaq_batch(
            close, 
            (11, 41, 10),      
            (2, 2, 0),         
            (30, 30, 0)        
        )
        
        
        assert isinstance(batch_result, dict)
        assert 'values' in batch_result
        assert 'periods' in batch_result
        assert 'fast_periods' in batch_result
        assert 'slow_periods' in batch_result
        
        
        assert batch_result['values'].shape[0] == 4
        assert batch_result['values'].shape[1] == len(close)
        
        
        
        
        assert batch_result['values'].shape == (4, len(close))


    def test_maaq_batch_with_metadata(self, test_data):
        """Test MAAQ batch computation with metadata"""
        close = np.array(test_data['close'], dtype=np.float64)
    
        
        batch_result = maaq_batch(
            close, 
            (11, 31, 10),      
            (2, 4, 2),         
            (25, 35, 10)       
        )
        
        
        periods = batch_result['periods']
        fast_periods = batch_result['fast_periods']
        slow_periods = batch_result['slow_periods']
        
        
        assert len(periods) == 12
        assert len(fast_periods) == 12
        assert len(slow_periods) == 12
        
        
        assert batch_result['values'].shape == (12, len(close))
        
        
        


    def test_maaq_batch_2d(self, test_data):
        """Test MAAQ batch computation with 2D output"""
        close = np.array(test_data['close'], dtype=np.float64)
    
        
        batch_result = maaq_batch(
            close, 
            (11, 31, 20),      
            (2, 3, 1),         
            (30, 30, 0)        
        )
        
        
        metadata = list(zip(
            batch_result['periods'], 
            batch_result['fast_periods'], 
            batch_result['slow_periods']
        ))
        
        
        assert metadata == [(11, 2, 30), (11, 3, 30), (31, 2, 30), (31, 3, 30)]
        
        
        assert batch_result['values'].shape == (4, len(close))
        
        
        
        assert batch_result['values'].shape == (4, len(close))


    def test_maaq_stream(self, test_data):
        """Test MAAQ streaming interface - mirrors check_maaq_streaming"""
        close = test_data['close']
    
        
        period = 11
        fast_period = 2
        slow_period = 30
        
        
        close_array = np.array(close, dtype=np.float64)
        batch_result = maaq(close_array, period, fast_period, slow_period)
        
        
        stream = MaaqStream(period, fast_period, slow_period)
        stream_results = []
        
        for price in close:
            result = stream.update(price)
            
            stream_results.append(result if result is not None else np.nan)
        
        
        assert len(batch_result) == len(stream_results)
        
        
        
        for i in range(period, len(batch_result)):
            
            if np.isnan(batch_result[i]) and np.isnan(stream_results[i]):
                continue
            
            assert_close(stream_results[i], batch_result[i], atol=1e-9, 
                        msg=f"Streaming mismatch at index {i}")


    def test_maaq_different_periods(self, test_data):
        """Test MAAQ with various period values"""
        close = np.array(test_data['close'], dtype=np.float64)
    
        
        test_cases = [
            (5, 2, 10),
            (10, 3, 20),
            (20, 5, 40),
            (50, 10, 100),
        ]
        
        for period, fast_p, slow_p in test_cases:
            result = maaq(close, period, fast_p, slow_p)
            assert len(result) == len(close)
            
            
            valid_count = np.sum(np.isfinite(result[period:]))
            assert valid_count > len(close) - period - 5, \
                f"Too many NaN values for params=({period}, {fast_p}, {slow_p})"


    def test_maaq_batch_performance(self, test_data):
        """Test that batch computation works correctly (performance is secondary)"""
        close = np.array(test_data['close'][:1000], dtype=np.float64)  
    
        
        batch_result = maaq_batch(
            close, 
            (10, 30, 10),      
            (2, 2, 0),         
            (25, 35, 5)        
        )
        
        
        assert batch_result['values'].shape == (9, len(close))
        
        
        
        assert batch_result['values'].shape == (9, len(close))


    def test_maaq_edge_cases(self):
        """Test MAAQ with edge case inputs"""
        
        data = np.arange(1.0, 101.0, dtype=np.float64)
        result = maaq(data, 10, 2, 20)
        assert len(result) == len(data)
        
        
        assert np.all(np.isfinite(result[10:]))
        
        
        data = np.array([1.0, 10.0, 1.0, 10.0, 1.0, 10.0, 1.0, 10.0] * 20, dtype=np.float64)
        result = maaq(data, 5, 2, 10)
        assert len(result) == len(data)
        
        
        data = np.array([5.0] * 100, dtype=np.float64)
        result = maaq(data, 10, 2, 20)
        assert len(result) == len(data)
        
        for i in range(20, len(result)):
            assert_close(result[i], 5.0, atol=1e-9, msg=f"Constant value failed at index {i}")


    def test_maaq_warmup_period(self, test_data):
        """Test that warmup period is correctly calculated
        
        MAAQ follows ALMA's warmup semantics: first (period-1) values are NaN
        to indicate insufficient data for calculation.
        """
        close = np.array(test_data['close'][:50], dtype=np.float64)
        
        test_cases = [
            (5, 2, 10),    
            (10, 3, 20),   
            (20, 5, 30),   
            (30, 10, 40),  
        ]
        
        for period, fast_p, slow_p in test_cases:
            result = maaq(close, period, fast_p, slow_p)
            
            
            for i in range(period - 1):
                assert np.isnan(result[i]), \
                    f"Expected NaN at warmup index {i} for period={period}, got {result[i]}"


    def test_maaq_consistency(self, test_data):
        """Test that MAAQ produces consistent results across multiple calls"""
        close = np.array(test_data['close'][:100], dtype=np.float64)
        
        result1 = maaq(close, 11, 2, 30)
        result2 = maaq(close, 11, 2, 30)
        
        assert_close(result1, result2, atol=1e-15, msg="MAAQ results not consistent")


    def test_maaq_step_precision(self):
        """Test batch with very small step sizes"""
        data = np.arange(1, 51, dtype=np.float64)
    
        
        batch_result = maaq_batch(
            data, 
            (5, 7, 1),         
            (2, 3, 1),         
            (10, 10, 0)        
        )
        
        
        metadata = list(zip(
            batch_result['periods'], 
            batch_result['fast_periods'], 
            batch_result['slow_periods']
        ))
        
        assert metadata == [
            (5, 2, 10), (5, 3, 10),
            (6, 2, 10), (6, 3, 10),
            (7, 2, 10), (7, 3, 10)
        ]
        assert batch_result['values'].shape == (6, len(data))


    def test_maaq_batch_error_handling(self):
        """Test MAAQ batch error handling for edge cases"""
        
        empty = np.array([], dtype=np.float64)
        with pytest.raises(ValueError, match="maaq:"):
            maaq_batch(empty, (10, 20, 10), (2, 2, 0), (30, 30, 0))
        
        
        all_nan = np.full(100, np.nan, dtype=np.float64)
        with pytest.raises(ValueError, match="maaq:"):
            maaq_batch(all_nan, (10, 20, 10), (2, 2, 0), (30, 30, 0))
        
        
        small_data = np.array([1.0, 2.0, 3.0], dtype=np.float64)
        with pytest.raises(ValueError, match="maaq:"):
            maaq_batch(small_data, (5, 10, 5), (2, 2, 0), (30, 30, 0))
        
        
        data = np.random.randn(100).astype(np.float64)
        with pytest.raises(ValueError, match="maaq:"):
            maaq_batch(data, (200, 300, 50), (2, 2, 0), (30, 30, 0))


    def test_maaq_zero_copy_verification(self, test_data):
        """Verify MAAQ uses zero-copy operations"""
        
        close = np.array(test_data['close'][:100], dtype=np.float64)
        
        
        result = maaq(close, 11, 2, 30)
        assert len(result) == len(close)
        
        
        batch_result = maaq_batch(close, (10, 30, 10), (2, 2, 0), (25, 35, 5))
        assert batch_result['values'].shape[0] == 3 * 3  
        assert batch_result['values'].shape[1] == len(close)


    def test_maaq_stream_error_handling(self, test_data):
        """Test MAAQ stream error handling"""
        
        with pytest.raises(ValueError, match="maaq:"):
            MaaqStream(0, 2, 30)
        
        with pytest.raises(ValueError, match="maaq:"):
            MaaqStream(11, 0, 30)
        
        with pytest.raises(ValueError, match="maaq:"):
            MaaqStream(11, 2, 0)
        
        
        close = np.array(test_data['close'][:100], dtype=np.float64)
        period, fast_p, slow_p = 11, 2, 30
        
        batch_result = maaq(close, period, fast_p, slow_p)
        stream = MaaqStream(period, fast_p, slow_p)
        stream_results = []
        
        for price in close:
            result = stream.update(price)
            stream_results.append(result if result is not None else np.nan)
        
        
        for i in range(period, len(close)):
            if not np.isnan(batch_result[i]) and not np.isnan(stream_results[i]):
                assert_close(stream_results[i], batch_result[i], atol=1e-9,
                            msg=f"Stream mismatch at index {i}")


    def test_maaq_batch_warmup_consistency(self):
        """Test that batch warmup periods are consistent
        
        Each row in batch should have NaN values for the first (period-1) elements,
        matching ALMA's warmup semantics.
        """
        data = np.random.randn(50).astype(np.float64)
        
        result = maaq_batch(data, (5, 15, 5), (2, 2, 0), (10, 10, 0))
        
        
        
        
        for i, period in enumerate(result['periods']):
            row = result['values'][i]
            
            assert len(row) == len(data)
    
    
    
    
    def test_maaq_stream_reset(self):
        """Test MAAQ streaming reset functionality"""
        stream = MaaqStream(11, 2, 30)
        
        
        for i in range(20):
            stream.update(float(i))
        
        
        stream = MaaqStream(11, 2, 30)
        
        
        results = []
        for i in range(20):
            result = stream.update(float(i * 2))
            results.append(result if result is not None else np.nan)
        
        
        
        assert len(results) == 20
        
        
        for i in range(10, 20):
            assert not np.isnan(results[i]), f"Expected valid value at index {i}"
    
    def test_maaq_single_data_point(self):
        """Test MAAQ with single data point and period=1"""
        data = np.array([42.0], dtype=np.float64)
        
        
        
        with pytest.raises(BaseException):
            maaq(data, 1, 1, 1)


if __name__ == "__main__":
    
    print("Testing MAAQ module...")
    pytest.main([__file__, '-v'])
    print("MAAQ tests completed!")

File: tests/wasm/test_maaq.js

import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import {
    loadTestData,
    assertArrayClose,
    assertClose,
    isNaN,
    assertAllNaN,
    assertNoNaN,
    EXPECTED_OUTPUTS
} from './test_utils.js';
import { compareWithRust } from './rust-comparison.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

test.before(async () => {

    try {
        const wasmPath = path.join(__dirname, '../../pkg/vector_ta.js');
        const importPath = process.platform === 'win32'
            ? 'file:///' + wasmPath.replace(/\\/g, '/')
            : wasmPath;
        wasm = await import(importPath);
    } catch (error) {
        console.error('Failed to load WASM module. Run "wasm-pack build --features wasm --target nodejs" first');
        throw error;
    }

    testData = loadTestData();
});

test('MAAQ partial params', () => {

    const close = new Float64Array(testData.close);

    const result = wasm.maaq_js(close, 11, 2, 30);
    assert.strictEqual(result.length, close.length);
});

test('MAAQ accuracy', async () => {

    const close = new Float64Array(testData.close);

    const result = wasm.maaq_js(close, 11, 2, 30);

    assert.strictEqual(result.length, close.length);


    const expectedLastFive = [
        59747.657115949725,
        59740.803138018055,
        59724.24153333905,
        59720.60576365108,
        59673.9954445178,
    ];


    const last5 = result.slice(-5);
    assertArrayClose(
        last5,
        expectedLastFive,
        0.01,
        "MAAQ last 5 values mismatch"
    );


    await compareWithRust('maaq', result, 'close', {
        period: 11,
        fast_period: 2,
        slow_period: 30
    });
});

test('MAAQ default candles', async () => {

    const close = new Float64Array(testData.close);

    const result = wasm.maaq_js(close, 11, 2, 30);
    assert.strictEqual(result.length, close.length);


    await compareWithRust('maaq', result, 'close', {
        period: 11,
        fast_period: 2,
        slow_period: 30
    });
});

test('MAAQ zero period', () => {

    const inputData = new Float64Array([10.0, 20.0, 30.0]);


    assert.throws(() => {
        wasm.maaq_js(inputData, 0, 2, 30);
    });


    assert.throws(() => {
        wasm.maaq_js(inputData, 11, 0, 30);
    });


    assert.throws(() => {
        wasm.maaq_js(inputData, 11, 2, 0);
    });
});

test('MAAQ period exceeds length', () => {

    const dataSmall = new Float64Array([10.0, 20.0, 30.0]);

    assert.throws(() => {
        wasm.maaq_js(dataSmall, 10, 2, 30);
    });
});

test('MAAQ very small dataset', () => {

    const dataSingle = new Float64Array([42.0]);

    assert.throws(() => {
        wasm.maaq_js(dataSingle, 11, 2, 30);
    });
});

test('MAAQ empty input', () => {

    const dataEmpty = new Float64Array([]);

    assert.throws(() => {
        wasm.maaq_js(dataEmpty, 11, 2, 30);
    });
});

test('MAAQ all NaN', () => {

    const data = new Float64Array([NaN, NaN, NaN, NaN, NaN]);

    assert.throws(() => {
        wasm.maaq_js(data, 3, 2, 5);
    });
});

test('MAAQ reinput', () => {

    const close = new Float64Array(testData.close);


    const firstResult = wasm.maaq_js(close, 11, 2, 30);


    const secondResult = wasm.maaq_js(firstResult, 20, 3, 25);

    assert.strictEqual(secondResult.length, firstResult.length);



    for (let i = 40; i < secondResult.length; i++) {
        assert(isFinite(secondResult[i]), `NaN found at index ${i}`);
    }
});

test('MAAQ NaN handling', () => {

    const close = new Float64Array(testData.close);
    const period = 11;

    const result = wasm.maaq_js(close, period, 2, 30);

    assert.strictEqual(result.length, close.length);


    for (let i = 0; i < period - 1; i++) {
        assert(isNaN(result[i]), `Expected NaN at warmup index ${i}, got ${result[i]}`);
    }



    if (result.length > 240) {
        for (let i = 240; i < result.length; i++) {
            assert(!isNaN(result[i]), `Found unexpected NaN at index ${i}`);
        }
    }
});

test('MAAQ batch', () => {

    const close = new Float64Array(testData.close);


    const batch_result = wasm.maaq_batch_js(
        close,
        {
            period_range: [11, 41, 10],
            fast_period_range: [2, 2, 0],
            slow_period_range: [30, 30, 0]
        }
    );
    const metadata = wasm.maaq_batch_metadata_js(
        11, 41, 10,
        2, 2, 0,
        30, 30, 0
    );



    assert.strictEqual(metadata.length, 12);

    assert.strictEqual(metadata[0], 11);
    assert.strictEqual(metadata[1], 2);
    assert.strictEqual(metadata[2], 30);


    assert.strictEqual(batch_result.length, 4 * close.length);


    let row_idx = 0;
    for (let p = 11; p <= 41; p += 10) {
        const individual_result = wasm.maaq_js(close, p, 2, 30);


        const row_start = row_idx * close.length;
        const row = batch_result.slice(row_start, row_start + close.length);

        assertArrayClose(row, individual_result, 1e-9, `Period ${p}`);
        row_idx++;
    }
});

test('MAAQ different periods', () => {

    const close = new Float64Array(testData.close);


    const testCases = [
        [5, 2, 10],
        [10, 3, 20],
        [20, 5, 40],
        [50, 10, 100],
    ];

    for (const [period, fast_p, slow_p] of testCases) {
        const result = wasm.maaq_js(close, period, fast_p, slow_p);
        assert.strictEqual(result.length, close.length);


        let validCount = 0;
        for (let i = period; i < result.length; i++) {
            if (!isNaN(result[i])) validCount++;
        }
        assert(validCount > close.length - period - 5,
            `Too many NaN values for params=(${period}, ${fast_p}, ${slow_p})`);
    }
});

test('MAAQ batch performance', () => {

    const close = new Float64Array(testData.close.slice(0, 1000));


    const startBatch = performance.now();
    const batchResult = wasm.maaq_batch_js(
        close,
        {
            period_range: [10, 30, 10],
            fast_period_range: [2, 2, 0],
            slow_period_range: [25, 35, 5]
        }
    );
    const batchTime = performance.now() - startBatch;

    const startSingle = performance.now();
    const singleResults = [];
    for (let p = 10; p <= 30; p += 10) {
        for (let s = 25; s <= 35; s += 5) {
            singleResults.push(...wasm.maaq_js(close, p, 2, s));
        }
    }
    const singleTime = performance.now() - startSingle;


    console.log(`Batch time: ${batchTime.toFixed(2)}ms, Single time: ${singleTime.toFixed(2)}ms`);


    assertArrayClose(batchResult, singleResults, 1e-9, 'Batch vs single results');
});

test('MAAQ edge cases', () => {



    const data = new Float64Array(100);
    for (let i = 0; i < 100; i++) {
        data[i] = i + 1;
    }
    const result = wasm.maaq_js(data, 10, 2, 20);
    assert.strictEqual(result.length, data.length);


    for (let i = 10; i < result.length; i++) {
        assert(isFinite(result[i]), `NaN at index ${i}`);
    }


    const constantData = new Float64Array(100).fill(5.0);
    const constantResult = wasm.maaq_js(constantData, 10, 2, 20);
    assert.strictEqual(constantResult.length, constantData.length);


    for (let i = 20; i < constantResult.length; i++) {
        assertClose(constantResult[i], 5.0, 1e-9, `Constant prediction failed at index ${i}`);
    }
});

test('MAAQ batch metadata', () => {

    const metadata = wasm.maaq_batch_metadata_js(
        11, 31, 10,
        2, 4, 2,
        25, 35, 10
    );


    assert.strictEqual(metadata.length, 36);


    assert.strictEqual(metadata[0], 11);
    assert.strictEqual(metadata[1], 2);
    assert.strictEqual(metadata[2], 25);


    assert.strictEqual(metadata[3], 11);
    assert.strictEqual(metadata[4], 2);
    assert.strictEqual(metadata[5], 35);
});

test('MAAQ warmup period calculation', () => {


    const close = new Float64Array(testData.close.slice(0, 50));

    const testCases = [
        { period: 5, fast_p: 2, slow_p: 10 },
        { period: 10, fast_p: 3, slow_p: 20 },
        { period: 20, fast_p: 5, slow_p: 30 },
        { period: 30, fast_p: 10, slow_p: 40 },
    ];

    for (const { period, fast_p, slow_p } of testCases) {
        const result = wasm.maaq_js(close, period, fast_p, slow_p);


        for (let i = 0; i < period - 1 && i < result.length; i++) {
            assert(isNaN(result[i]),
                `Expected NaN at warmup index ${i} for period=${period}, got ${result[i]}`);
        }


        if (period - 1 < result.length) {
            assert(!isNaN(result[period - 1]),
                `Expected valid value at index ${period - 1} for period=${period}`);
        }
    }
});

test('MAAQ consistency across calls', () => {

    const close = new Float64Array(testData.close.slice(0, 100));

    const result1 = wasm.maaq_js(close, 11, 2, 30);
    const result2 = wasm.maaq_js(close, 11, 2, 30);

    assertArrayClose(result1, result2, 1e-15, "MAAQ results not consistent");
});

test('MAAQ parameter step precision', () => {

    const data = new Float64Array(50);
    for (let i = 0; i < 50; i++) {
        data[i] = i + 1;
    }

    const batch_result = wasm.maaq_batch_js(
        data,
        {
            period_range: [5, 7, 1],
            fast_period_range: [2, 3, 1],
            slow_period_range: [10, 10, 0]
        }
    );


    assert.strictEqual(batch_result.length, 6 * data.length);


    const metadata = wasm.maaq_batch_metadata_js(5, 7, 1, 2, 3, 1, 10, 10, 0);
    assert.strictEqual(metadata.length, 18);
});

test('MAAQ streaming simulation', () => {

    const close = new Float64Array(testData.close.slice(0, 100));
    const period = 11;
    const fast_period = 2;
    const slow_period = 30;


    const batchResult = wasm.maaq_js(close, period, fast_period, slow_period);


    assert.strictEqual(batchResult.length, close.length);


    for (let i = 0; i < period - 1; i++) {
        assert(isNaN(batchResult[i]), `Expected NaN at warmup index ${i}`);
    }


    let hasDifferentValues = false;
    for (let i = period - 1; i < close.length; i++) {
        assert(!isNaN(batchResult[i]), `Unexpected NaN at index ${i}`);
        if (Math.abs(batchResult[i] - close[i]) > 1e-9) {
            hasDifferentValues = true;
        }
    }
    assert(hasDifferentValues, "MAAQ should produce smoothed values after warmup");
});

test('MAAQ large period', () => {

    const data = new Float64Array(100);
    for (let i = 0; i < 100; i++) {
        data[i] = Math.sin(i * 0.1) * 100 + 1000;
    }

    const period = 50;
    const result = wasm.maaq_js(data, period, 5, 60);
    assert.strictEqual(result.length, data.length);


    for (let i = 0; i < period - 1; i++) {
        assert(isNaN(result[i]), `Expected NaN at warmup index ${i}`);
    }


    for (let i = period - 1; i < result.length; i++) {
        assert(isFinite(result[i]), `Expected finite value at index ${i}`);
    }
});




test('MAAQ batch with invalid ranges', () => {

    const data = new Float64Array(50);
    for (let i = 0; i < 50; i++) {
        data[i] = i + 1;
    }


    const desc = wasm.maaq_batch_js(data, {
        period_range: [20, 10, 5],
        fast_period_range: [2, 2, 0],
        slow_period_range: [30, 30, 0]
    });
    assert.strictEqual(desc.length, 3 * data.length, 'Expected 3 period combos');


    assert.throws(() => {
        wasm.maaq_batch_js(data, {
            period_range: [100, 200, 50],
            fast_period_range: [2, 2, 0],
            slow_period_range: [30, 30, 0]
        });
    });
});

test('MAAQ accuracy with expected values', () => {

    const close = new Float64Array(testData.close);
    const expected = EXPECTED_OUTPUTS.maaq;

    const result = wasm.maaq_js(
        close,
        expected.defaultParams.period,
        expected.defaultParams.fast_period,
        expected.defaultParams.slow_period
    );


    const last5 = result.slice(-5);
    assertArrayClose(
        last5,
        expected.last5Values,
        0.01,
        "MAAQ last 5 values mismatch with expected"
    );


    const warmupEnd = expected.defaultParams.period - 1;
    for (let i = 0; i < warmupEnd; i++) {
        assert(isNaN(result[i]), `Expected NaN at warmup index ${i}`);
    }
});

test('MAAQ single value with period 1', () => {

    const data = new Float64Array([42.0]);


    assert.throws(() => {
        wasm.maaq_js(data, 1, 1, 1);
    });
});
